enhancements/ingress/set-replace-append-delete-deny-headers.md

```diff
@@ -0,0 +1,491 @@
+---
+title: set-replace-append-delete-deny-headers
+authors:
+  - "@miheer"
+reviewers:
+  - "@frobware"
+  - "@gcs278"
+  - "@Miciah"
+approvers:
+  - "@frobware"
+  - "@gcs278"
+  - "@Miciah"
+api-approvers:
+  - "@JoelSpeed"
+  - "@deads2k"
+  - "@frobware"
+  - "@gcs278"
+  - "@Miciah"
+creation-date: 2022-12-02
+last-updated: 2022-12-14
+tracking-link:
+  - "https://issues.redhat.com/browse/NE-982"
+  - "https://issues.redhat.com/browse/RFE-464"
+status: TODO
+see-also:
+  - ""
+replaces:
+superseded-by:
+---
+
+# Set, Delete, Deny HTTP Response headers
+
+## Summary
+
+This enhancement extends the IngressController API and Route API to allow the user to set or delete
+HTTP response headers. Using the new API, users may specify a list of
+HTTP response headers either via IngressController API or Route API that should be set or deleted.
+Also, this enhancement creates an immutable(hard-coded) list of HTTP response headers which can't be set or deleted.
+
+## Motivation
+
+Certain HTTP response headers may need to be set, deleted for compliance reasons or other reasons. In OpenShift 3, cluster 
+administrators could use a custom HAProxy configuration template to set or delete headers, but using a custom
+template is not permitted in OpenShift 4.
+Certain HTTP response headers like HSTS (Strict-Transport-Security) need to be denied as we already provide a way to set them
+via route annotation.
```

Comment:
Would it not be better to provide a concrete API and deprecate the annotation API?

Reply:
Deprecating will be big job to do for now. We can try to achieve this in the next release.

Reply:
I wonder is this still true, given this is going into 4.14, do we want to think about this deprecation now?

Reply:
The HSTS annotation predates our era of learning that annotations were a mistake (3.7 I think https://docs.openshift.com/container-platform/3.7/admin_guide/managing_networking.html#admin-guide-enabling-hsts).  Since then we learned that the migration path from "temporary" annotation to formal schema basically never happens because never have the appetite to break consumers.

For types we own the schema for, annotation-based APIs should never be accepted.  Annotations
1. suffer from squatting: people can set the value before we add validation
2. suffer from lack of discoverability: we don't have a schema to inform generic editors
3. suffer from limitations of validation: certain kinds of validators aren't allowed on annotations, so hooks are more frequently used instead
4. suffer from low expressibility: we cannot extend future fields under a parent

For types we do not own schemas for, we can approve annotation-based APIs, but only when we can provide clear feedback for failures to the parties that can fix the problems.